Implementation of Switch component

benja opened the pull request:
React implementation of the Fabric Switch component.

I would appreciate some help in regards to accessibility. My assumption would be that it should be tabable, as well as toggleable with the spacebar.

pearofducks commented:
We need to define a focus state for this component (with Adi's help), but I think we get the a11y "for free" via the input/label elements? I'm able to navigate to the component via keyboard-nav and (de)activate it.

benja commented:
> We need to define a focus state for this component (with Adi's help), but I think we get at least some of the a11y "for free" via the input/label elements? I'm able to navigate to the component via keyboard-nav and (de)activate it.

I agree with you that it needs a focus state. I'm able to tab to the component (no visual focus though), but not toggle the component's value with my keyboard. Perhaps because it's an `onClick` and not `onChange`?

benja commented:
#13 solved the keyboard toggling for me. Does it still work as expected for you?

benja commented:
bcd4aac solved the keyboard toggling for me. Does it still work as expected for you?

tor-martin-storsletten commented:
Looks like accessible name (label text) is missing. The switch role requires an accessible name.
Unfortunately, an ARIA switch does not get accessible name from a native label element via the "for" attribute either (tested in Chrome with JAWS just for fun), so that's also a no-go.
Maybe a solution could be to use a button element instead of input/label? The button could wrap the component's children (which would act as the text content) alongside a span element that visually indicates the toggle state?

benja commented:
Thank you, @tor-martin-storsletten. I've now added support for the aria-labels and changed the input to a button for semantic accessibility.

I also went ahead and applied Tailwind's focus styles, @pearofducks, which we can change anytime. E.g. after consulting Adi.

pearofducks commented:
Sounds good!

alii commented on packages/box/src/component.tsx:
```diff
@@ -11,17 +11,26 @@ const setup = ({
+              // The Box element can't be a button or link in case someone puts an interactive element inside the box, which would result in invalid HTML and severe a11y issues.
+              if (
+                  typeof attrs.onClick === 'function' &&
+                  (event.keyCode === 13 || event.keyCode === 32)
```
`event.keyCode` is generally considered deprecated, you should look at using `event.key` instead.

alii reviewed (changes requested):
Minor suggestion

digitalsadhu commented on packages/switch/src/component.tsx:
```diff
@@ -0,0 +1,53 @@
+                className={classNames(
+                    [
+                        c.label,
+                        'focus:outline-none focus:ring ring-offset-1 ring-blue-200 rounded-full ',
```
( Super minor thing that you can completely choose to ignore 😆 ) I almost prefer to use a single object for everything and set to true rather than end up with multiple params like this:

```js
{
  c.label: true,
  'focus:outline-none focus:ring ring-offset-1 ring-blue-200 rounded-full ': true,
  [c.labelDisabled]: disabled,
}
```

benja replied:
I suppose it'd be more readable to those who haven't used classNames much before either. Although, I chose to put them in the first param as they will always return true, so mainly for abstraction and reducing redundancy.

Another way to make it shorter and more readable (for both approaches) would be to define a `const switchFocus = 'focus:outline-none focus:ring ring-offset-1 ring-blue-200 rounded-full';` and use that instead.

**Option 1 — two params**
```JS
className={classNames([c.label, switchFocus], {
    [c.labelDisabled]: disabled,
})}
```

**Option 2 — one param**

```JS
className={classNames({
    switchFocus: true,
    [c.label]: true,
    [c.labelDisabled]: disabled,
})}
```

Either one is fine with me, although I do prefer the former.

benja replied:
I suppose it'd be more readable to those who haven't used classNames much before either. I put them in the first param mainly for reducing redundancy as they will always return true.

Another way to make it shorter and more readable (for both approaches) would be to define a `const switchFocus = 'focus:outline-none focus:ring ring-offset-1 ring-blue-200 rounded-full';` and use that instead.

**Option 1 — two params**
```JS
className={classNames([c.label, switchFocus], {
    [c.labelDisabled]: disabled,
})}
```

**Option 2 — one param**

```JS
className={classNames({
    switchFocus: true,
    [c.label]: true,
    [c.labelDisabled]: disabled,
})}
```

Either one is fine with me, although I do prefer the former.

digitalsadhu commented on packages/switch/src/props.tsx:
```diff
@@ -0,0 +1,31 @@
+export interface SwitchProps {
```
interface works in all the same cases as type? Works in the mdx prop table etc?

benja replied:
Yeah it works. I usually prefer interfaces for props and `type`, `enum` etc for individual types used in interfaces.

digitalsadhu commented on packages/switch/src/props.tsx:
```diff
@@ -0,0 +1,31 @@
+    /**
+     * Whether the Switch is disabled
+     */
+    disabled?: boolean;
```
Where did we land with `disabled` @Skadefryd @pearofducks ? In the case of a switch is disabled still kinda maybe sorta an anti pattern or not so much?

pearofducks replied:
Yeah a disabled switch should be "no staahp" IMO

benja replied:
One use-case I can see for disabled switches would be when they require further verification or the setting is accessed behind a paywall. Although, an alternative to both of these cases could be to completely remove the disabled state and rather don't update the value of the switch on click and show a modal saying further verification, payment, etc is required to complete action instead.

tor-martin-storsletten replied:
I would recommend a modal instead of disabled state. Disabled state is an antipattern, as there will always be users who don't understand why the element is disabled and that may cause cognitive stress, which nobody likes.

benja replied:
I have removed the disabled property and added a message in the docs advising the developer to use a modal or toast instead.

digitalsadhu reviewed (approved):
Lookin good!
